### PR TITLE
[modules] Update ADC temperature to be configurable

### PR DESCRIPTION
**Pull request by fvantienen**
Make temperature module through ADC more configurable.

### CODE REVIEW

File: sw/airborne/modules/sensors/temp_adc.c [hunk 1]
```diff
@@ -81,19 +83,25 @@ static struct adc_buf temp_buf3;
...
+static float calc_ntc(int16_t raw_adc, uint16_t pull_up_r, float a, float b, float c)
 {
-  float temp_c;
-  //calc for NTC
-  temp_c = log(((10240000 / raw_temp) - 10000) * 10);
-  //temp_c = 1/(0.001129148+(0.000234125*temp_c)+(0.0000000876741*temp_c*temp_c*temp_c));
-  temp_c = 1 / (0.000603985662844 + (0.000229995493730 * temp_c) + (0.000000067653027 * temp_c *
-                temp_c * temp_c));
-  temp_c = temp_c - 273.15; //convert do celcius
+#ifdef TEMP_ADC_12BIT
```
Comment (flixr) [inline sw/airborne/modules/sensors/temp_adc.c]: What is this define about?

Comment (fvantienen) [inline sw/airborne/modules/sensors/temp_adc.c]: If you have 12 bits ADC versus a 10 bits ADC.
Originally this was written for a 10 bits, so I added this define to support 12 bits as well.

**Comment by fvantienen** [comment]
By the way the usage of this module will be in different pull request, but I'm currently busy getting everything to master etc.

File: sw/airborne/modules/sensors/temp_adc.c (hunk 1 above)
Comment (flixr) [inline sw/airborne/modules/sensors/temp_adc.c]: Then this should at least be documented, or even better automatically take the correct value depending on what ADC resolution your mcu has.

Comment (fvantienen) [inline sw/airborne/modules/sensors/temp_adc.c]: I will add some documentation, bit in my board header file I set the correct value ;)

**Comment by gautierhattenberger** [comment]
I have never really looked at this module, but I think it is not really properly named. It only supports 2 type of sensor and it is not even possible to configure for a random sensor (like with an offset and sensitivity).
What do you think about supporting a generic sensor, or change the module name ?

**Comment by fvantienen** [comment]
It supports a generic sensor, namely NTC. This is the default temperature sensor used.
Note that the NTC is highly non-linear, that's why you cannot find your simple offset or sensitivity. You instead have an Alpha and Beta value from the datasheet.

File: sw/airborne/modules/sensors/temp_adc.h [hunk 2]
```diff
@@ -28,12 +28,6 @@
 
 #include "std.h"
 
-/// flag to enable sending every new measurement via telemetry
-bool temp_adc_sync_send;
```
Comment (flixr) [inline sw/airborne/modules/sensors/temp_adc.h]: why remove this? IMHO it should rather be added to settings in the module xml

Comment (fvantienen) [inline sw/airborne/modules/sensors/temp_adc.h]: Because in my opinion this was only added for debugging and instead I made the define for that.

Comment (flixr) [inline sw/airborne/modules/sensors/temp_adc.h]: I think it's meant to allow enabling/disabling the reporting at runtime, which IMHO is quite useful.

**Comment by fvantienen** [comment]
I updated everything as requested, is it ok to merge?

File: sw/airborne/modules/sensors/temp_adc.c [hunk 3]
```diff
@@ -31,7 +31,8 @@
 #include "subsystems/datalink/downlink.h"
 #include BOARD_CONFIG
 
-float temp_c1, temp_c2, temp_c3;
+bool temp_adc_sync_send = false;
```
Comment (flixr) [inline sw/airborne/modules/sensors/temp_adc.c]: Should be initialized with the param TEMP_ADC_SYNC_SEND

**Comment by flixr** [comment]
The doc for TEMP_ADC_12BIT was accidentally removed again?
Which boards have 10 bit? AFIAK the stm32 boards have all 12bit, right?
So it would be good to default to the correct one.

File: sw/airborne/modules/sensors/temp_adc.c (hunk 3 above)
Comment (fvantienen) [inline sw/airborne/modules/sensors/temp_adc.c]: That is done in the initialization function.

**Comment by fvantienen** [comment]
Fixed the default to be for STM32

File: sw/airborne/modules/sensors/temp_adc.c (hunk 3 above)
Comment (flixr) [inline sw/airborne/modules/sensors/temp_adc.c]: sorry, missed that...

**Comment by flixr** [comment]
It would be better to default to 12bit if STM32 and to 10 bit otherwise (for lpc21), e.g.
```
#ifndef ADC_RESOLUTION
#if defined(STM32F1) || defined(STM32F4)
#define ADC_RESOLUTION 4096
#else
#define ADC_RESOLUTION 1024
#endif
#endif
```

**Comment by fvantienen** [comment]
Ok I added this to `adc.h` because that seemed appropriate.

**Comment by gautierhattenberger** [comment]
Even better could be to have this in adc_arch.h files, but I guess it can be done later.

**Comment by gautierhattenberger** [comment]
Side question out of curiosity: in which application are you using these sensors ? meteo or something else ?

**Comment by fvantienen** [comment]
We are using it for measuring the temperature of the main Motor and the Battery.

**Pull request merged**